File: src/platform/ios/timer.m
//
//  timer.m
//  baye-ios
//
//  Created by loong on 15/8/15.
//
//

#import <Foundation/Foundation.h>

static NSTimer* _timer1;
NSTimeInterval _timerInterval;

#define ratio 100.0

@interface GamTimerRecv: NSObject
@property void(*callback)();
@end

@implementation GamTimerRecv
- (void)timerFired
{
    if (_callback) {
        _callback();
    }
}
@end

static GamTimerRecv *receiver;

int gam_timer_interval()
{
    return (int)(_timerInterval * ratio);
}

void gam_timer_init()
{
    receiver = [GamTimerRecv new];
}

void gam_timer_set_callback(void(*cb)())
{
    receiver.callback = cb;
}

void gam_timer_open(int interval)
{
    _timerInterval = interval / ratio;
    dispatch_async(dispatch_get_main_queue(), ^{
        if (!_timer1) {
            _timer1 = [NSTimer scheduledTimerWithTimeInterval:_timerInterval
                                                       target:receiver
                                                     selector:@selector(timerFired)
                                                     userInfo:nil repeats:YES];
        }
    });
}

void gam_timer_close()
{
    dispatch_async(dispatch_get_main_queue(), ^{
        if (_timer1) {
            [_timer1 invalidate];
            _timer1 = nil;
        }
    });
}


static NSTimer* timer2;
static GamTimerRecv *receiver2;

void gam_timer2_open(int interval, void(*callback))
{
    receiver2 = [GamTimerRecv new];
    receiver2.callback = callback;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (!timer2) {
            timer2 = [NSTimer scheduledTimerWithTimeInterval:interval / ratio
                                                       target:receiver2
                                                     selector:@selector(timerFired)
                                                     userInfo:nil repeats:YES];
        }
    });
}

File: src/platform/js/timer.c
//
//  timer.c
//  baye-ios
//
//  Created by loong on 15/8/15.
//
//
#include <emscripten.h>

#define NULL ((void*)0)
#define ratio 10.0

typedef void(*timer_cb)(void);

typedef struct {
    int started;
    int active;
    int interval;
    timer_cb cb;
} timer_t;

static timer_t timer1 = {
    0, 0, 1, NULL
};

static void schedule_timer(timer_t* timer);

void _timer_cb(void*t) {
    timer_t* timer = (timer_t*)t;
    if (timer->active && timer->cb) {
        timer->cb();
    }
    schedule_timer(timer);
}

static void schedule_timer(timer_t* timer) {
    emscripten_async_call(_timer_cb, timer, timer->interval * ratio);
}

int gam_timer_interval()
{
    return timer1.interval;
}

void gam_timer_init()
{
}

void gam_timer_set_callback(void(*cb)(void))
{
    timer1.cb = cb;
}

static void _timer_open(timer_t* timer) {
    timer->active = 1;
    if (!timer->started) {
        timer->started = 1;
        schedule_timer(timer);
    }
}

void gam_timer_open(int interval)
{
    timer1.interval = interval;
    _timer_open(&timer1);
}

void gam_timer_close()
{
    timer1.active = 0;
}


static timer_t timer2 = {
    0, 0, 1, NULL
};

void gam_timer2_open(int interval, void(*callback))
{
    timer2.interval = interval;
    timer2.cb = callback;
    _timer_open(&timer2);
}